### ISSUE TITLE
An API documentation question

### ISSUE THREAD

**Issue by geoffmcl** [htacg/tidy-html5#248]
I was setting about generating the API documentation, but have a **BIG** problem, question!

At present we only have **one** site for the API documentation, namely http://www.htacg.org/tidy-html5/tidylib_api/, which is presently version 4.9.35.

We might keep that, but now it seems we have the need to add two or more sites, like -

 - Documentation 4.9.35 - already done, but may need moving.
 - Documentation 5.0.0
 - Documentation 5.1.8
 - Others, later, onward...
 
Obviously for the most part they will be the **same** but there are some important changes already, like say the two(2) new features 5.1.8 offers of `--vertical-space auto`, and `--indent-with-tabs yes|no` not present in 5.0.0.

So if I choose to download the 5.0.0 `release/5.0.0` branch then I would like to refer to `5.0.0` API docs. And if I choose to go with a later stable release, say tag `5.1.8`, then I want to be reading `5.1.8` API docs.

At first I thought of somehow combining them...

But since this documentation is generated using doxygen I do not think there is a way to sort of combine them together, like the reference for the vertical space API - http://www.htacg.org/tidy-html5/tidylib_api/quick_ref.html#vertical-space - be split into two parts - something like -

```
  ---------------------------------------------------------------------------
  <h4>vertical-space 5.0.0</h4>
  <tr><td>Type: Boolean</td></tr>
  <tr><td>Default: no</td></tr>
  <tr><td>Example: y/n, yes/no, t/f, true/false, 1/0</td></tr>
  <tr><td>This option specifies if Tidy should add some empty lines for readability.</td></tr>
  ---------------------------------------------------------------------------
  <h4>vertical-space 5.1.8</h4>
  <tr><td>Type: AutoBool</td></tr>
  <tr><td>Default: no</td></tr>
  <tr><td>Example: auto, y/n, yes/no, t/f, true/false, 1/0</td></tr>
  <tr><td>This option specifies if Tidy should add some extra empty lines for 
    readability. Default is 'no'. If set to 'auto', will eliminate nearly 
   all newline chars.</td></tr>
  ---------------------------------------------------------------------------
```

But that can not work, since the document generator, doxygen, must sort of have access to two branches in this case, `release/5.0.0` and `master`, tag 5.1.8 at the **same** time to split it when the API have changed. No, that seems almost impossible...

My next idea is that the current http://www.htacg.org/tidy-html5/tidylib_api/ leads only to a **version** index, which directs into say sub-directories like -

```
<a href="http://www.htacg.org/tidy-html5/tidylib_api/4.9.35">Documentation 4.9.35</a><br>
<a href="http://www.htacg.org/tidy-html5/tidylib_api/5.0.0">Documentation 5.0.0</a><br>
<a href="http://www.htacg.org/tidy-html5/tidylib_api/5.1.8">Documentation 5.1.8</a><br>
... and so on ...
```

where each sub-directory is a **full** set of API docs for that version... each auto generated with the repo set at that version/branch/tag...

This would work, but I would really need help to set that up... at this time I only have a very vague idea about how yaml voodoo works to generate this site... and quite minimal understanding of doxygen generation...

But how do other projects, which have a changing API per version, do this?

Well it seems others like FLTK also separate it into a version index, and have a different URL per version -

```
FLTK 2.0 -> http://fltk.org/doc-2.0/html/
FLTK 1.3.3 -> http://fltk.org/doc-1.3/
FLTK 1.1.10 -> http://www.fltk.org/documentation.php/doc-1.1/toc.html
```

Likewise for CMAKE -

```
CMake 3.3.1 -> http://www.cmake.org/cmake/help/v3.3/
CMake 3.2.3 -> http://www.cmake.org/cmake/help/v3.2/
...
CMake 2.8.12 -> http://www.cmake.org/cmake/help/v2.8.12/cmake.html
```

So it seems there is a lots of precedence in using site sub-directories to hold each version... and that makes sense... 

Once a particular version generation is done, and uploaded, it will stay like that forever... at least while that version is still available in distribution...

So that is sort of settled, use sub-directories for storage, unless others have other ideas...

But as already stated, I will really need **BIG HELP** with that...

**Comment by geoffmcl** [comment]
Hi Denis,

Thanks for replying. I am adding your emails here. It is better done here rather than direct emails to me, so that others can see them, and comment...

On 09/09/15 00:31, Denis Denisov wrote:

- https://coderwall.com/p/vemncg/what-is-the-difference-rest-docutils-sphinx-readthedocs

I recommend to look at Readthedocs.org

Similar: http://pythonhosted.org/.

readthedocs, aka rtfd / rtd / readthedocs.org is a website for serving
documentation for software projects.

It builds and hosts sphinx documentation projects.

Each software project's documentation may have it's own .rst files,
sphinx extensions and sphinx theme.

Sincerely,
Denis Denisov

And then added -

http://breathe.readthedocs.org/en/latest/

But I was **not** really looking for an `alternate` documentation generation system, but someone interested in taking over the task of document generation and pushing it to the site...

Naturally that person, or persons, could choose other than the xsltproc/doxygen tool chain presently used... that could be discussed and agreed...

The important thing is to generate the API docs for each version, like -

```
<a href="tidylib_api/4.9.35">4.9.35</a><br>
<a href="tidylib_api/5.0.0">5.0.0</a><br>
<a href="tidylib_api/5.1.9">5.1.9</a><br>
... and so on ...
```

Maybe this could be setup on one of the public CI (continuous integration) servers, like as @pedromorgan once suggested #173, #175, and @daaain $16, ... using https://travis-ci.org/ or http://jenkins.freeflightsim.org/, or where-ever... 

Pete, are you still interested in this? We want it now ;=))

##### So again, anyone out there interested in helping with this API document generation side of the project?

**Comment by denji** [comment]
> The important thing is to generate the API docs for each version, like -

Actually the readthedocs implementation is based on the ability:
```
http://html5-tidy.readthedocs.org/en/latest/
http://html5-tidy.readthedocs.org/en/stable/
http://html5-tidy.readthedocs.org/en/dev/
http://html5-tidy.readthedocs.org/en/{version}/
Downloads support: PDF, ePub, HTML
Quick open Github editor
````
- https://docs.readthedocs.org/en/latest/badges.html#status-badges
- https://docs.readthedocs.org/en/latest/alternate_domains.html
- https://docs.readthedocs.org/en/latest/webhooks.html
- https://docs.readthedocs.org/en/latest/localization.html

**Comment by pedromorgan** [comment]
readthe docs wont work, cos it uses sphinx and if we want to document the "code", then really its doxygen.. 

I've got a site up and running previous with the api docs, and the idea of automating it..
http://tidy5docs-htacg.rhcloud.com/

However I do not think there is value in maintains more then one version.. instead I suggest to use terms such as 
"Introduced in version x.y" or "depreciated iny.x " and alike...

Also i dont think there is a lot of value in the "webiste" and would rooll the "website" into docygen and one repos and onne plase for the whole lot... imho

**Comment by geoffmcl** [comment]
@pedromorgan thanks for the comment on `readthedocs`... it seems ok with me that we stick to doxygen... which is already setup...

If the version feature changes were as simple as just two ideas - (a) introduced in v.x.y and (b) deprecated in v.y.z - then maybe it **could** all be covered in terms... could still try that...

And yes, http://tidy5docs-htacg.rhcloud.com/ looks good. I note it is presently 4.9.20... How can I `download` those site files? I do not see any login, admin, etc... but maybe I have forgotten...

And then how do I go about updating that to say the [5.0.0 release](https://github.com/htacg/tidy-html5/releases/tag/5.0.0)? And again getting all the site files...

This seems like a good start ;=))

**Comment by pedromorgan** [comment]
What's you "openshift.com" login (email) so I can add u to openshift
project...

Re updating, I;ll make a pull request to update..
It will user fabric, which is a python lib and makes scripting a doddle..

U can (or i can) then setup this up on a cron somewhere to autoupdate (same
as FG docs)

Pete

**Comment by pedromorgan** [comment]
Ohh, and can u sent me a couple of new/depreceated functions so I can show
the markup??

pete

**Comment by geoffmcl** [comment]
@pedromorgan, well my opeshift.com email is the usual ubuntu _at_ geoffair _dot_ info I think...

Re: updating to 5.0.0

This sounds good, but you did **not** answer `how do I download the site pages`?

My idea at this time is the have `that` site always with the **latest**... Maybe triggered when the version.txt is updated... but for now just up to 5.0.0...

But for the http://www.html-tidy.org/tidylib_api/ I am still thinking of the sub-directory versioning mentioned above... so it is **important** I collect the 5.0.0 page **before** it is updated to the next version, probably 5.1.9...

Running out of time tonight, but two options spring to mind - `vertical-space` went from Bool to AutoBool with the new `auto`, and `indent-with-tabs` new to 5.1, not in 5.0 - IIRC ;=))

**Comment by denji** [comment]
readthedocs for me easier to make changes will be harder to anyone, but it's up to you guys.

**Comment by pedromorgan** [comment]
Hi Denis...
The snag is that doxygen created the "C api docs"... and sphinx dont so
that,, well it does.. but c coders hate it.. so not an option... basically..

I use readthedocs a lot, so understand, but for this project.. its doxygen..
If there was a "readthedocX".. then i would be pig in shit.. would run down
the street naked to the pub.... ;-))

**Comment by vielmetti** [comment]
I have to say that `doxygen` is a big system and pulls in a lot of dependencies. I'm building the whole shebang on a Raspberry Pi 2 and it's pulling in most of a gigabyte of code just to build the docs (and all I really need is a little man page to build correctly).

**Comment by pedromorgan** [comment]
@vielmetti Yep doxgen does...

Its wicked.. and can eat c++, md, rst et all..

its mature and .. ahem.. am a dox.dev.tester .. 
see http://api-docs.freeflightsim.org/
Whicih is doxygen of doxyies.. .,. or pixies.. must be stoened !! ??????

The Doxygen docs are creates by "doxxygen parsing the files", and not compiling..
Similar in php where comments required..

Compiling the FG docs .. particular the OSG part takes around 20 mins,,
and had to be compiled twice in a chicken and egg situation.., 
hence first run is placing eggs
when next run is identical, but can sit on first egg.. 
so indeed for our purpose.. we crate the egg first, and then the chicken..

Anyway.. we want some one stop cool resourse to eat the html5 api, and how to use it ;-)))

**Comment by balthisar** [comment]
With the addition of api.html-tidy.org, can we close this issue?

**Comment by geoffmcl** [comment]
@balthisar absolutely! [api.html-tidy.org](http://api.html-tidy.org) looks great... thanks...
